### ISSUE TITLE
nodetool refresh is inefficient 

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#4223]
Usually when doing nodetool refresh, a user is bringing SSTables from a foreign cluster.
Such clusters usually will have different token ranges and not all data will be used in the new node.

Nodetool refresh, when called, immediately triggers a reshard, which will scan all the data - including the data that doesn't belong to the node, and reshard them to Scylla CPUs. The user then has to call another, manual nodetool cleanup.

This is very inefficient for the following reasons:
- Refresh is writing data that does not have a place in this node. This is 1 useless write, but that will also trigger another read on those bytes later when the actual cleanup is called
- For users calling nodetool cleanup in a sequence (a common pattern when adding many source nodes to one destination node), we will be forced to cleanup SSTables that are already cleaned up.

In my view, it would be much better to allow the resharding that is triggered from refresh (if not all cleanups) to have cleanup capabilities as well: not only we never write bytes that don't belong in the first place, the need to call cleanup at each step is no more making for an overall more efficient import.

**Comment by avikivity** [comment]
Even better is to have refresh (which should be renamed to import) send non-local data to a replica that owns it.

**Comment by glommer** [comment]
Maybe we should have a specialized API for that?
My fear is users with old procedures doing a n x m import and ending up with thousands of copies flying around over the network.

My preference would be: 
- fix refresh to at least not be that inefficient
- Add an import which on top of everything streams data. Then you don't have to do n x m imports in that new API.

**Comment by avikivity** [comment]
Sure. Merging clean into refresh is not too hard, and also fixes a data resurrection bug (in case that node later regains a token range that was populated by refresh).

But note that resharding is done on an as-needed basis, so it will need to change to always make a reshard+clean pass.

We should record vnode token ranges in Scylla.db (like we record shard ranges) so we have more information about the sstable. Likely we can reuse the same data structure, just be more careful.

**Comment by glommer** [comment]
resharding is called on an as-needed basis, yes, but it will always be called for nodetool refresh (unless of course the SSTables are already sharded).

If the SSTables are perfectly sharded already in this node, doesn't that mean that it necessarily came from that node? (thinking...)

**Comment by avikivity** [comment]
No. Consider a single-shard node. Any sstable is perfectly sharded for that node.

**Comment by avikivity** [comment]
It could have come from any node that has nr_shards that is a power-of-two smaller than this_node_nr_shards (including exactly equal),  and sstable that has just one partition will also be perfectly sharded.

**Comment by avikivity** [comment]
Of course, that partition is probably 112GB in size.

**Comment by glommer** [comment]
true.

If we have a reshard+cleanup operation, we can just call it unconditionally upon refresh. We lose a bit of efficiency in the cases where everything was already perfectly clean and sharded and coming from the same node

**Comment by avikivity** [comment]
Like restore from backup? Sad

**Comment by slivne** [comment]
@asias didn't you fix this

**Comment by asias** [comment]
> @asias didn't you fix this

If user uses the new load and stream version of refresh, we do not have to run cleanup manually. My patch does not fix the old  refresh method.
